chatgpt-shell-ollama.el

```diff
@@ -84,6 +86,65 @@ VALIDATE-COMMAND handler."
          :token-width 4
          :context-window 32768)))
 
+(defun chatgpt-shell-ollama--fetch-model-versions ()
+  (mapcar (lambda (model)
+            (string-remove-suffix ":latest" (map-elt model 'name)))
+          (map-elt (shell-maker--json-parse-string
+                    (map-elt (shell-maker-make-http-request
+                              :async nil
+                              :url (concat chatgpt-shell-ollama-api-url-base "/api/tags"))
+                             :output))
+                   'models)))
+
+(defun chatgpt-shell-ollama--parse-token-width (quantization)
+  (string-match "^[FQ]\\([1-9][0-9]*\\)" quantization)
+  (string-to-number (match-string 1 quantization)))
+
+(defun chatgpt-shell-ollama--make-model (version)
+  (let* ((data (shell-maker--json-parse-string
+                (map-elt (shell-maker-make-http-request
+                          :async nil
+                          :url (concat chatgpt-shell-ollama-api-url-base "/api/show")
+                          :data `((model . ,version)))
+                         :output)))
+         (token-width (chatgpt-shell-ollama--parse-token-width
+                       (map-elt (map-elt data 'details) 'quantization_level)))
+         (context-window (cdr (cl-find-if (lambda (cell)
+                                            (string-suffix-p "context_length" (symbol-name (car cell))))
+                                          (map-elt data 'model_info)))))
+    (chatgpt-shell-ollama-make-model
+     :version version
+     :token-width token-width
+     :context-window context-window)))
+
+(cl-defun chatgpt-shell-ollama-load-models (&key override)
+  "Query ollama for the locally installed models and add them to
+`chatgpt-shell-models' unless a model with the same name is
+already present. If OVERRIDE is non-nil (interactively with a
+prefix argument), remove all ollama models from
+`chatgpt-shell-models' before inserting locally installed models."
+  (interactive (list :override current-prefix-arg))
+  (let* ((ollama-predicate (lambda (model)
+                             (string= (map-elt model :provider) "Ollama")))
+         ;; Find the index of the first ollama model so that the new ones will
+         ;; be placed in the same part of the list.
+         (ollama-index (or (cl-position-if ollama-predicate chatgpt-shell-models)
+                           (length chatgpt-shell-models))))
+    (when override
```

Comment:
Oooh, your override implementation is much nicer than I had in mind. I was thinking of just overriding the entire list (wiping non-Ollama models) without considering merging. I like your merge strategy more.

How about we combine both approaches and:

:override t = override entire list (wipe non-Ollama models)
:override nil = merge Ollama models.

ps. Not sure if you've seen, but we have `M-x chatgpt-shell-reload-default-models`, which is handy for re-instating default models.